<a href="https://colab.research.google.com/github/HenriqueCCdA/bootCampAluraDataScience/blob/master/modulo4/extra/ICU.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 100)

# 2) Explorando os dados

# 2.1) Importando os dados

In [ ]:
path = "https://github.com/HenriqueCCdA/bootCampAluraDataScience/blob/master/modulo4/dados/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx?raw=true"

dados = pd.read_excel(path)
dados.head(n=5)

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

In [ ]:
nl, nc = dados.shape
print(f"Numero de linhas : {nl} ")
print(f"Numero de colunas: {nc} ")

Numero de linhas : 1925 
Numero de colunas: 231 


In [ ]:
print(dados.dtypes)

PATIENT_VISIT_IDENTIFIER        int64
AGE_ABOVE65                     int64
AGE_PERCENTIL                  object
GENDER                          int64
DISEASE GROUPING 1            float64
                               ...   
RESPIRATORY_RATE_DIFF_REL     float64
TEMPERATURE_DIFF_REL          float64
OXYGEN_SATURATION_DIFF_REL    float64
WINDOW                         object
ICU                             int64
Length: 231, dtype: object


In [ ]:
for name in dados.columns:
  if 'OXYGEN_SATURATION' in name:
    print(name)

OXYGEN_SATURATION_MEAN
OXYGEN_SATURATION_MEDIAN
OXYGEN_SATURATION_MIN
OXYGEN_SATURATION_MAX
OXYGEN_SATURATION_DIFF
OXYGEN_SATURATION_DIFF_REL


In [ ]:
dados.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 229 columns]

In [ ]:
dados.iloc[:21,[0,-2,-1]]

PATIENT_VISIT_IDENTIFIER    WINDOW  ICU
0                          0       0-2    0
1                          0       2-4    0
2                          0       4-6    0
3                          0      6-12    0
4                          0  ABOVE_12    1
5                          1       0-2    1
6                          1       2-4    1
7                          1       4-6    1
8                          1      6-12    1
9                          1  ABOVE_12    1
10                         2       0-2    0
11                         2       2-4    0
12                         2       4-6    0
13                         2      6-12    0
14                         2  ABOVE_12    1
15                         3       0-2    0
16                         3       2-4    0
17                         3       4-6    0
18                         3      6-12    0
19                         3  ABOVE_12    0
20                         4       0-2    0

In [ ]:
n_pacientes_inicial = len(dados['PATIENT_VISIT_IDENTIFIER'].unique())
print('Numero de pacientes', n_pacientes_inicial)

Numero de pacientes 385


In [ ]:
dados['AGE_PERCENTIL'].unique()

array(['60th', '90th', '10th', '40th', '70th', '20th', '50th', '80th',
       '30th', 'Above 90th'], dtype=object)

In [ ]:
names = ['HTN', 'IMMUNOCOMPROMISED', 'OTHER', 'GENDER', 'AGE_ABOVE65',
         'DISEASE GROUPING 1', 'DISEASE GROUPING 2', 'DISEASE GROUPING 3',
         'DISEASE GROUPING 4', 'DISEASE GROUPING 5', 'DISEASE GROUPING 6',
         'WINDOW', 'ICU', 'AGE_PERCENTIL']
for name in names:
  print(f'{name:20} ->', dados[name].unique())

HTN                  -> [ 0.  1. nan]
IMMUNOCOMPROMISED    -> [ 0.  1. nan]
OTHER                -> [ 1.  0. nan]
GENDER               -> [0 1]
AGE_ABOVE65          -> [1 0]
DISEASE GROUPING 1   -> [ 0.  1. nan]
DISEASE GROUPING 2   -> [ 0.  1. nan]
DISEASE GROUPING 3   -> [ 0.  1. nan]
DISEASE GROUPING 4   -> [ 0.  1. nan]
DISEASE GROUPING 5   -> [ 1.  0. nan]
DISEASE GROUPING 6   -> [ 1.  0. nan]
WINDOW               -> ['0-2' '2-4' '4-6' '6-12' 'ABOVE_12']
ICU                  -> [0 1]
AGE_PERCENTIL        -> ['60th' '90th' '10th' '40th' '70th' '20th' '50th' '80th' '30th'
 'Above 90th']


## 2.2) Resumo:

>Geral:
* Número de linhas  : 1925 
* Número de colunas : 231
* Colunas 0, 1, 3 e 230 tipo int64
* Colunas 2 e 229 tipo string
* Colunas 4 a 228 tipo float64
* Número de pacientes: 385 (cada paciente tem mutiplicas entradas por janelas de tempo)

>Colunas categoricas 0 ou 1:
* HTN                  -> [ 0.  1. nan]
* IMMUNOCOMPROMISED    -> [ 0.  1. nan]
* OTHER                -> [ 1.  0. nan]
* GENDER               -> [0 1]
* AGE_ABOVE65          -> [1 0]
* DISEASE GROUPING 1   -> [ 0.  1. nan]
* DISEASE GROUPING 2   -> [ 0.  1. nan]
* DISEASE GROUPING 3   -> [ 0.  1. nan]
* DISEASE GROUPING 4   -> [ 0.  1. nan]
* DISEASE GROUPING 5   -> [ 1.  0. nan]
* DISEASE GROUPING 6   -> [ 1.  0. nan]

>Colunas categoricas Multi-Label:
* WINDOW               -> ['0-2' '2-4' '4-6' '6-12' 'ABOVE_12']
* AGE_PERCENTIL        -> ['60th' '90th' '10th' '40th' '70th' '20th' '50th' '80th' '30th'
 'Above 90th']

>Possivel divisão das idades:
* 10th - 0  a 10 anos
* 20th - 11 a 20 anos
* 30th - 21 a 30 anos
* 40th - 31 a 20 anos
* 50th - 41 a 50 anos
* 60th - 51 a 60 anos
* 70th - 71 a 80 anos
* 80th - 81 a 90 anos
* Above 90th - 91 a 1.. anos

>OBS:
* Hypertension (HTN) 
* Imunodepressivos (IMMUNOCOMPROMISED)  
* UTI (intensive care unit)

> Nas colunas 13 até 228 as variaveis são continuas
>
>Por exemplo:\
OXYGEN_SATURATION_MEDIAN - mediana \
OXYGEN_SATURATION_MEAN   - media\
OXYGEN_SATURATION_MIN    - minimo\
OXYGEN_SATURATION_MAX    - maximo\
OXYGEN_SATURATION_DIFF   - max - min         
OXYGEN_SATURATION_DIFF_REL  - diff/median - (max - min)/median  



# 3) Tratamento dos dados

In [ ]:
def pacientes_por_janela_ICU(dados):
  '''
  ------------------------------------------------------------------------------
  Função que conta se o paciente esta ou não na UTI para cada janela
  ------------------------------------------------------------------------------
  @param dados - dateFrame
  ------------------------------------------------------------------------------
  @return ICU_window_0_2  - numero de pacientes que foram para UTI janela 0-2
  @return ICU_window_2_4  - numero de pacientes que foram para UTI janela 2-4
  @return ICU_window_4_6  - numero de pacientes que foram para UTI janela 4-6
  @return ICU_window_6_12 - numero de pacientes que foram para UTI janela 6-12
  ------------------------------------------------------------------------------
  '''

  ICU_window_0_2 = dados.query("WINDOW == '0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER']
  ICU_window_2_4 = dados.query("WINDOW == '2-4' and ICU==1")['PATIENT_VISIT_IDENTIFIER']
  ICU_window_4_6 = dados.query("WINDOW == '4-6' and ICU==1")['PATIENT_VISIT_IDENTIFIER']
  ICU_window_6_12 = dados.query("WINDOW == '6-12' and ICU==1")['PATIENT_VISIT_IDENTIFIER']

  print('Numero de ICU igual a 1')
  print('Janela 0-2 :' ,len(ICU_window_0_2))
  print('Janela 2-4 :' ,len(ICU_window_2_4))
  print('Janela 4-6 :' ,len(ICU_window_4_6))
  print('Janela 6-12:',len(ICU_window_6_12))

  return ICU_window_0_2, ICU_window_2_4, ICU_window_4_6, ICU_window_6_12

In [ ]:
dados.iloc[:21,[0,-2,-1]]

PATIENT_VISIT_IDENTIFIER    WINDOW  ICU
0                          0       0-2    0
1                          0       2-4    0
2                          0       4-6    0
3                          0      6-12    0
4                          0  ABOVE_12    1
5                          1       0-2    1
6                          1       2-4    1
7                          1       4-6    1
8                          1      6-12    1
9                          1  ABOVE_12    1
10                         2       0-2    0
11                         2       2-4    0
12                         2       4-6    0
13                         2      6-12    0
14                         2  ABOVE_12    1
15                         3       0-2    0
16                         3       2-4    0
17                         3       4-6    0
18                         3      6-12    0
19                         3  ABOVE_12    0
20                         4       0-2    0

In [ ]:
ICU_window_0_2, _, _, _ = pacientes_por_janela_ICU(dados)

Numero de ICU igual a 1
Janela 0-2 : 32
Janela 2-4 : 59
Janela 4-6 : 99
Janela 6-12: 130


In [ ]:
ICU_window_0_2 # pacientes que precisam ser retirados

5         1
85       17
135      27
205      41
235      47
240      48
270      54
310      62
425      85
450      90
520     104
530     106
545     109
585     117
720     144
880     176
1010    202
1030    206
1130    226
1165    233
1195    239
1305    261
1350    270
1410    282
1430    286
1530    306
1575    315
1645    329
1795    359
1810    362
1830    366
1895    379
Name: PATIENT_VISIT_IDENTIFIER, dtype: int64

## 3.1) Definido as funções que faram o tratamento

In [ ]:
def retira_paciente_primeira_janela(matriz):
  '''
  ------------------------------------------------------------------------------
  Retira da base de dados os pacientes que foram para UTI na pimeira janela
  0-2
  ------------------------------------------------------------------------------
  @param matriz - dados brutos
  ------------------------------------------------------------------------------
  @return matriz sem o pacientes que que foram para UTI na pimeira janela
  ------------------------------------------------------------------------------
  '''

  window = matriz.query("WINDOW=='0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER']

  matriz = matriz.query("PATIENT_VISIT_IDENTIFIER not in @window")

  return matriz

def submatriz_preenchimento(submatriz):
  '''
  ------------------------------------------------------------------------------
  Preenche as variaveis continuas utilizando. Esta função trabalho com submatriz
  da matriz principal
  ------------------------------------------------------------------------------
  @param submatriz - submatriz agrupapor paciente e UCIs
  ------------------------------------------------------------------------------
  @return submatriz preenchida quando 
  ------------------------------------------------------------------------------
  '''
  
  submatriz_var_continuas = submatriz.iloc[:, 13:-2]
  submatriz_var_categorica_inicio = submatriz.iloc[:, :13]
  submatriz_var_categorica_final  = submatriz.iloc[:,-2:]

  # preenchendo os valores NaN 

  submatriz_var_continuas = submatriz_var_continuas.fillna(method='ffill')\
                                                   .fillna(method='bfill')

  submatriz_preenchida = pd.concat([submatriz_var_categorica_inicio, 
                                      submatriz_var_continuas,  
                                      submatriz_var_categorica_final],
                                      ignore_index=False, axis=1)
  
  return submatriz_preenchida


def preenchendo_var_continuas(matriz):

  '''
  ------------------------------------------------------------------------------
  Preenche as variaveis continuas utilizando os somente os dados para quando 
  ITU = 0. 
  ------------------------------------------------------------------------------
  @param matriz - dados brutos
  ------------------------------------------------------------------------------
  @return matriz sem o pacientes que que foram para UTI na pimeira janela
  ------------------------------------------------------------------------------
  '''

  matriz_preenchida = matriz.groupby(['PATIENT_VISIT_IDENTIFIER', 'ICU'],
                                     as_index=False).apply(submatriz_preenchimento)
  matriz_preenchida = matriz_preenchida.reset_index().drop(['level_0', 
                                                            'level_1'], axis=1)

  return matriz_preenchida


def uma_linha_por_paciente(submatriz):
  '''
  ------------------------------------------------------------------------------
  Reduz todas as janelas do paciente a apenas um linha. Os valores são pegos
  da janela 0-2
  ------------------------------------------------------------------------------
  @param submatriz - Submatriz agrupa por pacientes
  ------------------------------------------------------------------------------
  @return Retorna os valores da janela 0-2 com a informa se o paciente foi ou 
  não para a UTI
  ------------------------------------------------------------------------------
  '''

  if np.any(submatriz['ICU']):
    submatriz.iloc[:,-1] = 1

  return submatriz.iloc[0,:]

## 3.2) Manipulando os dados

* Retirando os pacientes que foram para a UTI na janela 0-2
* Preenchimento dos valores NaN nas variaveis continuas
* retirandos os valores NaN que restaram

In [ ]:
dados_tratados = dados.copy()
# Retirando os pacientes que foram para a UTI na janela 0-2
dados_tratados = retira_paciente_primeira_janela(dados_tratados)
# Preenchimento dos valores NaN nas variaveis continuas
dados_tratados = preenchendo_var_continuas(dados_tratados)
# retira todas alinhas que ainda tem valores NaN
dados_tratados = dados_tratados.dropna()

In [ ]:
dados_tratados.head(n=15)

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                          0            1  ...       0-2    0
1                          0            1  ...       2-4    0
2                          0            1  ...       4-6    0
3                          0            1  ...      6-12    0
4                          0            1  ...  ABOVE_12    1
5                          2            0  ...       0-2    0
6                          2            0  ...       2-4    0
7                          2            0  ...       4-6    0
8                          2            0  ...      6-12    0
9                          2            0  ...  ABOVE_12    1
10                         3            0  ...       0-2    0
11                         3            0  ...       2-4    0
12                         3            0  ...       4-6    0
13                         3            0  ...      6-12    0
14                         3            0  ...  ABOVE_12    0

[15 rows x 231 columns]

##3.2.1) Checando se os pacientes foram corretamente retirados

In [ ]:
_, _, _, _, = pacientes_por_janela_ICU(dados_tratados) # verificando se tem algum paciente ICU no intervalo 0-2

Numero de ICU igual a 1
Janela 0-2 : 0
Janela 2-4 : 27
Janela 4-6 : 67
Janela 6-12: 98


##3.2.2)  Checando se anda existem NaN

In [ ]:
for col in dados_tratados:
  if dados_tratados[col].isna().sum():
    print("Valores NaN na coluna", col, dados_tratados[col].isna().sum())

In [ ]:
nl, nc = dados_tratados.shape
print(f"Numero de linhas : {nl} ")
print(f"Numero de colunas: {nc} ")

Numero de linhas : 1645 
Numero de colunas: 231 


In [ ]:
dados_tratados['AGE_PERCENTIL'].unique()

array(['60th', '10th', '40th', '70th', '20th', '50th', '80th', '30th',
       '90th', 'Above 90th'], dtype=object)

## 3.3) Modificando a coluna AGE_PERCENTIL

AGE_PERCENTIL| valor proposto |
:-----------:|:--------------:|
   10th      |   -1.00  |
   20th      |   -0.75  |          
   30th      |   -0.50  |     
   40th      |   -0.25  |     
   50th      |    0.00  | 
   60th      |    0.25  |       
   70th      |    0.50  | 
   80th      |    0.75  |
   Above 90th|    1.00  |          


In [ ]:
map = { '10th'      : -1.0,
        '20th'      : -0.75,
        '30th'      : -0.5,
        '40th'      : -0.25,
        '50th'      :  0.0,
        '60th'      :  0.25,
        '70th'      :  0.5,
        '80th'      :  0.75,
        'Above 90th':  1.0,
} 

In [ ]:
dados_tratados['AGE_PERCENTIL'] = dados_tratados['AGE_PERCENTIL'].map(map)
dados_tratados['AGE_PERCENTIL'].dtype # checando o tipo

In [ ]:
dados_tratados.head(n=15)

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                          0            1  ...       0-2    0
1                          0            1  ...       2-4    0
2                          0            1  ...       4-6    0
3                          0            1  ...      6-12    0
4                          0            1  ...  ABOVE_12    1
5                          2            0  ...       0-2    0
6                          2            0  ...       2-4    0
7                          2            0  ...       4-6    0
8                          2            0  ...      6-12    0
9                          2            0  ...  ABOVE_12    1
10                         3            0  ...       0-2    0
11                         3            0  ...       2-4    0
12                         3            0  ...       4-6    0
13                         3            0  ...      6-12    0
14                         3            0  ...  ABOVE_12    0

[15 rows x 231 columns]

In [ ]:
dados_tratados_por_paciente = dados_tratados.groupby('PATIENT_VISIT_IDENTIFIER', as_index=False).apply(uma_linha_por_paciente)

In [ ]:
dados_tratados_por_paciente.head()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...  WINDOW  ICU
0                         0            1  ...     0-2    1
1                         2            0  ...     0-2    1
2                         3            0  ...     0-2    0
3                         4            0  ...     0-2    0
4                         5            0  ...     0-2    0

[5 rows x 231 columns]

In [ ]:
n_pacientes, _ = dados_tratados_por_paciente.shape
print(f"Numero de pacientes inicial : {n_pacientes_inicial}")
print(f"Numero de pacientes validos : {n_pacientes} ")

Numero de pacientes inicial : 385
Numero de pacientes validos : 351 


# 4) Salvando os dados tratados 

In [ ]:
dados_tratados_por_paciente.to_csv('dados_tratados_por_paciente.csv')